File: src/FullSpectrumWebForms_Diagnostic/wwwroot/js/diagnostic/controls/diagnosticManager.ts
/// <reference path="../../../../../fullSpectrumWebForms/wwwroot/js/core/controlBase.ts" />
/// <reference path="../../../../../fullSpectrumWebForms/wwwroot/js/controls/html/htmlControlBase.ts" />
/// <reference path="../../../../node_modules/@types/jquery/index.d.ts" />

namespace controls {

    export class diagnosticManager extends core.controlBase {

        initialize(type: string, index: number, id: string, properties: { property: string, value: any }[]) {
            super.initialize(type, index, id, properties);
        }

        checkIfElementExists(id: string) {
            let elem = $('#' + id);
            return elem.length != 0;
        }

        getElementTextFromId(id: string) {
            let elem = $('#' + id);
            if (elem.length == 0)
                return null;
            return elem.text();
        }

        getFSWControlText(id: string) {
            let control = core.manager.getControl(id) as html.htmlControlBase;
            if (control && control.element) {
                return control.element.text();
            }
            else
                return null;
        }
        getElementStyleFromId(parameters: { id: string, style: string }) {
            let elem = $('#' + parameters.id);
            if (elem.length == 0)
                return null;
            return elem.css(parameters.style);
        }

        getFSWControlStyle(parameters: { id: string, style: string }) {
            let control = core.manager.getControl(parameters.id) as html.htmlControlBase;
            if (control && control.element) {
                return control.element.css(parameters.style);
            }
            else
                return null;
        }


        getElementValFromId(id: string) {
            let elem = $('#' + id);
            if (elem.length == 0)
                return null;
            return elem.val();
        }

        getFSWControlVal(id: string) {
            let control = core.manager.getControl(id) as html.htmlControlBase;
            if (control && control.element)
                return control.element.val();
            return null;
        }

        setElementValFromId(parameters: {
            id: string,
            val: any
        }) {
            let elem = $('#' + parameters.id);
            if (elem.length != 0)
                elem.val(parameters.val);
            return null;
        }

        setFSWControlVal(parameters: {
            id: string,
            val: any
        }) {
            let control = core.manager.getControl(parameters.id) as html.htmlControlBase;
            if (control && control.element)
                control.element.val(parameters.val);
            return null;
        }

        sendKeysFromId(parameters: {
            id: string,
            key: any
        }) {
            let elem = $('#' + parameters.id);
            if (elem.length == 0)
                return null;

            (elem as any).sendkeys(parameters.key);
        }

        sendKeysFSWControl(parameters: {
            id: string,
            key: any
        }) {
            let control = core.manager.getControl(parameters.id) as html.htmlControlBase;
            if (control && control.element)
                (control.element as any).sendkeys(parameters.key);
            return null;
        }

        triggerElementValFromId(parameters: {
            id: string,
            ev: any
        }) {
            let elem = $('#' + parameters.id);
            if (elem.length == 0)
                return null;
            elem.trigger(parameters.ev);
        }

        triggerFSWControl(parameters: {
            id: string,
            ev: any
        }) {
            let control = core.manager.getControl(parameters.id) as html.htmlControlBase;
            if (control && control.element)
                control.element.trigger(parameters.ev);
            return null;
        }

        clickOnElementFromId(id: string) {
            let elem = $('#' + id);
            if (elem.length != 0)
                elem.trigger('click');
            return null;
        }

        clickOnElement(id: string) {
            let control = core.manager.getControl(id) as html.htmlControlBase;
            if (control && control.element)
                control.element.trigger('click');
            return null;
        }
        closeTab() {
            window.close();
        }
    }
}
core.controlTypes['DiagnosticManager'] = () => new controls.diagnosticManager();

File: src/FullSpectrumWebForms_Diagnostic/Controls/DiagnosticManager.cs
using FSW.Controls.Html;
using FSW.Core;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace FSW.Diagnostic.Controls
{
    public class DiagnosticManager : ControlBase
    {
        public override void InitializeProperties()
        {

        }

        public Task<bool> CheckIfElementExists(string id, bool autoLock = true)
        {
            return CallCustomClientEvent<bool>("checkIfElementExists", id, autoLock);
        }

        public Task<string> GetElementText(string id, bool autoLock = true)
        {
            return CallCustomClientEvent<string>("getElementTextFromId", id, autoLock);
        }
        public Task<string> GetElementText(HtmlControlBase control, bool autoLock = true)
        {
            return CallCustomClientEvent<string>("getFSWControlText", control.Id, autoLock);
        }
        public Task<string> GetElementStyle(string id, string style, bool autoLock = true)
        {
            return CallCustomClientEvent<string>("getElementStyleFromId", new
            {
                id,
                style
            }, autoLock);
        }
        public Task<string> GetElementStyle(HtmlControlBase control, string style, bool autoLock = true)
        {
            return CallCustomClientEvent<string>("getFSWControlStyle", new
            {
                id = control.Id,
                style
            }, autoLock);
        }

        public Task<T> GetElementVal<T>(string id, bool autoLock = true)
        {
            return CallCustomClientEvent<T>("getElementValFromId", id, autoLock);
        }
        public Task<T> GetElementVal<T>(HtmlControlBase control, bool autoLock = true)
        {
            return CallCustomClientEvent<T>("getFSWControlVal", control.Id, autoLock);
        }

        public Task SetElementVal<T>(string id, T val, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("setElementValFromId", new
            {
                id,
                val
            }, autoLock);
        }
        public Task SetElementVal<T>(HtmlControlBase control, T val, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("setFSWControlVal", new
            {
                id = control.Id,
                val
            }, autoLock);
        }

        public Task TriggerElement(string id, string eventToTrigger, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("triggerElementFromId", new
            {
                id,
                ev = eventToTrigger
            }, autoLock);
        }
        public Task TriggerElement(HtmlControlBase control, string eventToTrigger, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("triggerFSWControl", new
            {
                id = control.Id,
                ev = eventToTrigger
            }, autoLock);
        }

        public Task SendKeys(string id, string key, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("sendKeysFromId", new
            {
                id,
                key
            }, autoLock);
        }
        public Task SendKeys(HtmlControlBase control, string key, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("sendKeysFSWControl", new
            {
                id = control.Id,
                key
            }, autoLock);
        }

        public Task TriggerChange(string id, bool autoLock = true)
        {
            return TriggerElement(id, "change", autoLock);
        }
        public Task TriggerChange(HtmlControlBase control, bool autoLock = true)
        {
            return TriggerElement(control, "change", autoLock);
        }

        public Task TriggerEnter(string id, bool autoLock = true)
        {
            return SendKeys(id, "\n", autoLock);
        }
        public Task TriggerEnter(HtmlControlBase control, bool autoLock = true)
        {
            return SendKeys(control, "\n", autoLock);
        }

        public Task ClickOnElement(string id, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("clickOnElementFromId", id, autoLock);
        }
        public Task ClickOnElement(HtmlControlBase control, bool autoLock = true)
        {
            return CallCustomClientEvent<object>("clickOnElement", control.Id, autoLock);
        }

        private Task<T> CallCustomClientEvent<T>(string name, object parameters, bool autoLock)
        {

            IDisposable lockObj = autoLock ? Page.ServerSideLock : null;
            try
            {
                return CallCustomClientEvent<T>(name, parameters);
            }
            finally
            {
                lockObj?.Dispose();
            }
        }

        public void CloseTab()
        {
            CallCustomClientEvent("closeTab");
        }

    }
}
